app/content/scripts/webviewPreload.js

```diff
@@ -165,6 +165,180 @@
+      var passwordElem = fields[1]
+
+      if (!passwordElem) {
+        return
+      }
+
+      if (!usernameElem) {
+        if (credentials[action]) {
+          credentials[action].push([passwordElem, usernameElem])
+        } else {
+          credentials[action] = [[passwordElem, usernameElem]]
+        }
+        // Ask the main process for the credentials
+        ipcRenderer.send('get-password', formOrigin, action)
+      } else {
+        // Wait for user to pick the username, then autofill password
+        usernameElem.addEventListener('focus', e => {
+          let rect = usernameElem.getBoundingClientRect()
+          ipcRenderer.send('show-username-list', formOrigin, action, {
+            bottom: rect.bottom,
+            left: rect.left,
+            width: rect.width
+          })
+        })
+      }
+
+      // Whenever a form is submitted, offer to save it in the password manager
+      // if the credentials have changed.
+      form.addEventListener('submit', e => {
+        var fields = getFormFields(form, true)
+        // Re-get action in case it has changed
+        var action = form.action || document.location.href
+        var usernameElem = fields[0] || {}
+        var passwordElem = fields[1] || {}
+        ipcRenderer.send('save-password', usernameElem.value, passwordElem.value,
+                         formOrigin, action)
+      })
+    })
+
+    ipcRenderer.on('got-password', (e, username, password, origin, action) => {
+      // TODO: If there are multiple accounts, this always autofills the
+      // most recently-added one. We should detect when the username is
+      // entered and get the corresponding password.
+      var elems = credentials[action]
+      if (formOrigin === origin && elems) {
+        elems.forEach((elem) => {
+          // Autofill password
+          elem[0].value = password
+          if (username && elem[1]) {
+            // Autofill the username if there is one
+            elem[1].value = username
+          }
+        })
+      }
+    })
+  })
+
+  /**
+   * Gets form fields.
+   * @param {Element} form - The form to inspect
+   * @param {boolean} isSubmission - Whether the form is being submitted
+   * @return {Array.<Element>}
+   */
+  function getFormFields (form, isSubmission) {
+    var passwords = getPasswordFields(form, isSubmission)
+
+    // We have no idea what is going on with a form that has 0 or >3 password fields
+    if (passwords.length === 0 || passwords.length > 3) {
+      return [null, null, null]
+    }
+
+    // look for any form field that has username-ish attributes
+    var username = form.querySelector(['input[type=text][autocomplete=email i]']) ||
+        form.querySelector(['input[type=text][autocomplete=username i]']) ||
+        form.querySelector(['input[type=text][name=email i]']) ||
+        form.querySelector(['input[type=text][name=username i]']) ||
+        form.querySelector(['input[type=text][name=user i]'])
```

Comment:
Does Firefox or Chrome look more extensively for an input without any of those attributes? Possibly by looking at siblings that come before the password page if none of these are found?